lib/rails_param/param.rb

```diff
@@ -25,35 +14,45 @@ def param!(name, type, options = {}, &block)
       return unless params.include?(name) || check_param_presence?(options[:default]) || options[:required]
 
       begin
-        params[name] = coerce(params[name], type, options)
+        parameter = RailsParam::Param::Parameter.new(
+          name: name,
+          value: params[name],
+          type: type,
+          options: options
+        )
+
+        # coerce value
+        parameter.value = coerce(
```

Comment:
I think it would make sense to do the coercion on the value first, and then initializing the `RailsParam::Param::Parameter` (insert "yo dawg" meme here) with the coerced value.

My rationale for that is you'll never initialize a parameter with a value that doesn't match its type.